signer/src/crypto/secret_uri.rs

```diff
@@ -0,0 +1,153 @@
+/// ```
+///
+/// Where:
+/// - 1 denotes a phrase or hex string. If this is not provided, the [`DEV_PHRASE`] is used
+///   instead.
+/// - 2's denote optional "derivation junctions" which are used to derive keys. Each of these is
+///   separated by "/". A derivation junction beginning with "/" (ie "//" in the original string)
+///   is a "hard" path.
+/// - 3 denotes an optional password which is used in conjunction with the phrase provided in 1
+///   to generate an initial key. If hex is provided for 1, it's ignored.
+///
+/// Notes:
+/// - If 1 is a `0x` prefixed 64-digit hex string, then we'll interpret it as hex, and treat the hex bytes
+///   as a seed/MiniSecretKey directly, ignoring any password.
+/// - Else if the phrase part is a valid BIP-39 phrase, we'll use the phrase (and password, if provided)
+///   to generate a seed/MiniSecretKey.
+/// - Uris like "//Alice" correspond to keys derived from a DEV_PHRASE, since no phrase part is given.
+///
+/// There is no correspondence mapping between `SURI` strings and the keys they represent.
+/// Two different non-identical strings can actually lead to the same secret being derived.
+/// Notably, integer junction indices may be legally prefixed with arbitrary number of zeros.
+/// Similarly an empty password (ending the `SURI` with `///`) is perfectly valid and will
+/// generally be equivalent to no password at all.
+///
+/// # Examples
+///
+/// Parse [`DEV_PHRASE`] secret URI with junction:
+///
+/// ```
+/// # use subxt_signer::{SecretUri, DeriveJunction, DEV_PHRASE, ExposeSecret};
+/// # use std::str::FromStr;
+/// let suri = SecretUri::from_str("//Alice").expect("Parse SURI");
+///
+/// assert_eq!(vec![DeriveJunction::from("Alice").harden()], suri.junctions);
+/// assert_eq!(DEV_PHRASE, suri.phrase.expose_secret());
+/// assert!(suri.password.is_none());
+/// ```
+///
+/// Parse [`DEV_PHRASE`] secret URI with junction and password:
+///
+/// ```
+/// # use subxt_signer::{SecretUri, DeriveJunction, DEV_PHRASE, ExposeSecret};
+/// # use std::str::FromStr;
+/// let suri = SecretUri::from_str("//Alice///SECRET_PASSWORD").expect("Parse SURI");
+///
+/// assert_eq!(vec![DeriveJunction::from("Alice").harden()], suri.junctions);
+/// assert_eq!(DEV_PHRASE, suri.phrase.expose_secret());
+/// assert_eq!("SECRET_PASSWORD", suri.password.unwrap().expose_secret());
+/// ```
+///
+/// Parse [`DEV_PHRASE`] secret URI with hex phrase and junction:
+///
+/// ```
+/// # use subxt_signer::{SecretUri, DeriveJunction, DEV_PHRASE, ExposeSecret};
+/// # use std::str::FromStr;
+/// let suri = SecretUri::from_str("0xe5be9a5092b81bca64be81d212e7f2f9eba183bb7a90954f7b76361f6edb5c0a//Alice").expect("Parse SURI");
+///
+/// assert_eq!(vec![DeriveJunction::from("Alice").harden()], suri.junctions);
+/// assert_eq!("0xe5be9a5092b81bca64be81d212e7f2f9eba183bb7a90954f7b76361f6edb5c0a", suri.phrase.expose_secret());
+/// assert!(suri.password.is_none());
+/// ```
+pub struct SecretUri {
```

Comment:
you could maybe state that this code is extracted from substrate or something

I'm a little scared of it to be honest but looks identical to `SecretUri` in substrate...

Reply:
Yeah; in the `mod.rs` of the crypto folder I say that everything in the crypto folder comes from sp_core::crypto :) (though with tiny tweaks or doc comment updates).

I didn't like bringing it in particularly either but it is just a straight copy of the logic! Probably I could test this in isolation better too to prove it aligns.